kafka/common/pkg/kafka/consumer_handler.go

```diff
@@ -63,33 +85,47 @@ func (consumer *saramaConsumerHandler) Cleanup(session sarama.ConsumerGroupSessi
 // ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
 func (consumer *saramaConsumerHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
 	consumer.logger.Info(fmt.Sprintf("Starting partition consumer, topic: %s, partition: %d, initialOffset: %d", claim.Topic(), claim.Partition(), claim.InitialOffset()))
+	defer consumer.logger.Info(fmt.Sprintf("Stopping partition consumer, topic: %s, partition: %d", claim.Topic(), claim.Partition()))
 
-	// NOTE:
-	// Do not move the code below to a goroutine.
-	// The `ConsumeClaim` itself is called within a goroutine, see:
-	// https://github.com/Shopify/sarama/blob/master/consumer_group.go#L27-L29
+	// NOTE: Do not move the code below to a goroutine. The `ConsumeClaim` itself is called within a goroutine.
 	for message := range claim.Messages() {
 		if ce := consumer.logger.Check(zap.DebugLevel, "debugging"); ce != nil {
 			consumer.logger.Debug("Message claimed", zap.String("topic", message.Topic), zap.Binary("value", message.Value))
+			consumer.logger.Debug("message", zap.String("topic", message.Topic), zap.Int32("partition", message.Partition), zap.Int64("Offset", message.Offset))
 		}
 
-		mustMark, err := consumer.handler.Handle(session.Context(), message)
-
+		err := consumer.handleWithExponentialBackOffRetries(session, message)
 		if err != nil {
-			consumer.logger.Info("Failure while handling a message", zap.String("topic", message.Topic), zap.Int32("partition", message.Partition), zap.Int64("offset", message.Offset), zap.Error(err))
-			consumer.errors <- err
+			consumer.errors <- PermanentError
 		}
-		if mustMark {
+	}
+
+	return nil
+}
+
+func (consumer *saramaConsumerHandler) handleWithExponentialBackOffRetries(
+	session sarama.ConsumerGroupSession,
+	message *sarama.ConsumerMessage,
+) error {
+	return wait.ExponentialBackoff(consumer.backoff, func() (bool, error) {
```

Comment:
we want to rely on the CloudEvent SDK to do the retry. Also look at the delivery spec for additional strategies.

Reply:
Not here, because we use bindings apis so we don't rely on client